msgraphcore/middleware/retry_middleware.py

```diff
@@ -0,0 +1,184 @@
+import datetime
+import sys
+import time
+from email.utils import parsedate_to_datetime
+
+from msgraphcore.middleware.middleware import BaseMiddleware
+from msgraphcore.middleware.options.middleware_control import middleware_control
+
+
+class RetryMiddleware(BaseMiddleware):
+    """
+    TransportAdapter that allows us to specify the
+    retry policy for all requests
+    """
+
+    DEFAULT_TOTAL_RETRIES = 3
+    MAX_TOTAL_RETRIES = 10
+    DEFAULT_BACKOFF_FACTOR = 0.1
+    DEFAULT_RETRY_TIME_LIMIT = 180
+    MAXIMUM_BACKOFF = 120
+    _DEFAULT_RETRY_CODES = {429, 503, 504}
+
+    def __init__(self, retry_configs=None):
+        super().__init__()
+        self.total_retries: int = retry_configs.pop('retry_total', self.DEFAULT_TOTAL_RETRIES)
+        self.backoff_factor: float = retry_configs.pop(
+            'retry_backoff_factor', self.DEFAULT_BACKOFF_FACTOR
+        )
+        self.backoff_max: int = retry_configs.pop('retry_backoff_max', self.MAXIMUM_BACKOFF)
+        self.timeout: int = retry_configs.pop('retry_time_limit', self.DEFAULT_RETRY_TIME_LIMIT)
+
+        status_codes: [int] = retry_configs.pop('retry_on_status_codes', [])
+
+        self._retry_on_status_codes = set(status_codes) | self._DEFAULT_RETRY_CODES
+        self._allowed_methods = frozenset(['HEAD', 'GET', 'PUT', 'DELETE', 'OPTIONS', 'TRACE'])
+        self._respect_retry_after_header = True
+        self._retry_count = 0
+
+    @classmethod
+    def disable_retries(cls):
+        """
+        Disable retry functionality by setting total number of retries to allow to zero
+        Retry total takes precedence over all other counts.
+        """
+        return cls(retry_configs={"retry_total": 0})
+
+    def configure_retry_settings(self):
```

Comment:
If I have three requests.

1. The first one doesn't specify any per request retry configs so it uses the default config
2. The second one uses per request config so it uses the new config
3. What config will the third request use? Will it use the default config or the new config from the second request?

Reply:
I'll add tests for this then resolve.

Reply:
Good catch. This was failing. I fixed the issue and added tests.